src/foam/nanos/auth/UserAndGroupAuthService.js

```diff
@@ -0,0 +1,609 @@
+        `);
+      }
+    }
+  ],
+
+  constants: [
+    {
+      name: 'PASSWORD_VALIDATE_REGEX',
+      type: 'String',
+      value: '^.{6,}$'
+    },
+    {
+      name: 'PASSWORD_VALIDATION_ERROR_MESSAGE',
+      type: 'String',
+      value: 'Password must be at least 6 characters long.'
+    }
+  ],
+
+  properties: [
+    {
+      class: 'foam.dao.DAOProperty',
+      name: 'userDAO',
+      javaFactory: 'return (DAO) getLocalUserDAO();'
+    },
+    {
+      class: 'foam.dao.DAOProperty',
+      name: 'groupDAO',
+      javaFactory: 'return (DAO) getLocalGroupDAO();'
+    },
+    {
+      class: 'foam.dao.DAOProperty',
+      name: 'sessionDAO',
+      javaFactory: 'return (DAO) getLocalSessionDAO();'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'start',
+      javaCode: `
+        getUserDAO();
+        getGroupDAO();
+        getSessionDAO();
+      `
+    },
+    {
+      name: 'getCurrentUser',
+      type: 'foam.nanos.auth.User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaThrows: ['foam.nanos.auth.AuthenticationException'],
+      javaCode: `
+        // fetch context and check if not null or user id is 0
+        Session session = x.get(Session.class);
+        if ( session == null || session.getUserId() == 0 ) {
+          throw new AuthenticationException("Not logged in");
+        }
+
+        // get user from session id
+        User user = (User) userDAO_.find(session.getUserId());
+        if ( user == null ) {
+          throw new AuthenticationException("User not found: " + session.getUserId());
+        }
+
+        // check if user enabled
+        if ( ! user.getEnabled() ) {
+          throw new AuthenticationException("User disabled");
+        }
+
+        // check if user login enabled
+        if ( ! user.getLoginEnabled() ) {
+          throw new AuthenticationException("Login disabled");
+        }
+
+        // check if user group enabled
+        Group group = (Group) groupDAO_.find(user.getGroup());
+        if ( group != null && ! group.getEnabled() ) {
+          throw new AuthenticationException("User group disabled");
+        }
+
+        // check for two-factor authentication
+        if ( user.getTwoFactorEnabled() && ! session.getContext().getBoolean("twoFactorSuccess") ) {
+          throw new AuthenticationException("User requires two-factor authentication");
+        }
+
+        return user;
+      `
+    },
+    {
+      name: 'generateChallenge',
+      documentation: `A challenge is generated from the userID provided. This is
+        saved in a LinkedHashMap with TTL of 5.`,
+      type: 'String',
+      args: [
+        {
+          class: 'Long',
+          name: 'userID'
+        }
+      ],
+      javaThrows: ['foam.nanos.auth.AuthenticationException'],
+      javaCode: `
+        throw new UnsupportedOperationException("Unsupported operation: generateChallenge");
```

Comment:
This seems odd. Do you want this class to be a decorator instead so anything it doesn't implement moves along to the delegate?

Reply:
Decorate around what? I think this should remain the base class. There is a ticket on the NP side to add a decorator around this.